### PR TITLE
Semantic tokens: add module name support and improve performance and accuracy by traversing the hieAst along with source code

### PR DESCRIPTION
**Pull request by soulomoon**
fix https://github.com/haskell/haskell-language-server/issues/3957

Things have been done:
1. Switch `Name` to `Identifier` in the implementation and add `ModuleName` to the `HsSemanticTokenType`
2. Strip ``` ` ` ``` and `()`, and split out qualified names. e.g.``` `Preclude.length` ``` to ```Preclude.```  `length`
3. add tokenizer to walk ast with the souce rope to get more accurate result and faster. Should fix https://github.com/haskell/haskell-language-server/issues/3983.
4. add type sig to semanticConfig's TH result

### CODE REVIEW

**Comment by michaelpj** [comment]
This looks like it's still a work in progress?

**Comment by soulomoon** [comment]
> This looks like it's still a work in progress?
Yes, I decicded to to the module name split in here a well.

**Comment by michaelpj** [comment]
Looks like we could also do `module <name> where` but maybe that's not in the HIE ast?

**Comment by michaelpj** [comment]
Okay, so the parentheses are only included in the span if it's an operator section. That seems fine. The question is whether we _should_ highlight the parens when it's an operator section.

**Comment by soulomoon** [comment]
Before we support overlapping tokens.

The following might be a good idea:
Drop the "``, ()" for now, as suggested by @konn https://github.com/haskell/haskell-language-server/issues/3969#issuecomment-1894718215.
We do not need to consider the special split involed "``, ()" for now.

**Comment by michaelpj** [comment]
But if we drop the parens for operator sections around qualified names, we should do it for normal operator sections like `(+)` too for consistency

**Comment by soulomoon** [comment]
Ready to merge.

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs [hunk 1]
```diff
@@ -5,12 +5,15 @@
...
 {-# LANGUAGE OverloadedLabels      #-}
+{-# LANGUAGE OverloadedRecordDot   #-}
 {-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE PolyKinds             #-}
 {-# LANGUAGE RecordWildCards       #-}
 {-# LANGUAGE ScopedTypeVariables   #-}
 {-# LANGUAGE TemplateHaskell       #-}
 {-# LANGUAGE TypeFamilies          #-}
 {-# LANGUAGE UnicodeSyntax         #-}
+{-# OPTIONS_GHC -Wno-missing-signatures #-}
```
Comment (michaelpj) [inline plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs:16]: do you want this? better to add the signatures maybe?

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Query.hs [hunk 2]
```diff
@@ -62,39 +70,43 @@ nameNameSemanticFromHie hieKind rm ns = do
...
       where
-        xs = mkNameSet $ rights $ M.keys $ M.filterWithKey inclusion bd
+        xs = S.fromList $ M.keys $ M.filterWithKey inclusion bd
         inclusion :: Identifier -> IdentifierDetails a -> Bool
         inclusion a b = not $ exclusion a b
         exclusion :: Identifier -> IdentifierDetails a -> Bool
         exclusion idt IdentifierDetails {identInfo = infos} = case idt of
-          Left _  -> True
+          Left _  -> False
           Right _ -> any isEvidenceContext (S.toList infos)
```
Comment (michaelpj) [inline plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Query.hs:99]: These cases could do with a comment explaining why we're excluding or not

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/SemanticConfig.hs [hunk 3]
```diff
@@ -34,6 +34,7 @@ docName tt = case tt of
     TTypeSynonym     -> "type synonyms"
     TTypeFamily      -> "type families"
     TRecordField     -> "record fields"
+    TModuleName      -> "module names"
```
Comment (michaelpj) [inline plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/SemanticConfig.hs:37]:
```suggestion
    TModuleName      -> "modules"
```
I think this is more consistent with the others

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Types.hs [hunk 4]
```diff
@@ -43,6 +44,7 @@ data HsSemanticTokenType
   | TTypeSynonym -- Type synonym
   | TTypeFamily -- type family
   | TRecordField -- from match bind
+  | TModuleName -- from match bind
```
Comment (michaelpj) [inline plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Types.hs:47]: comment is wrong
```suggestion
  | TModule -- from match bind
```

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Utils.hs [hunk 5]
```diff
@@ -99,3 +101,63 @@ showSpan x = show (srcSpanStartLine x) <> ":" <> show (srcSpanStartCol x) <> "-"
...
+
+-- nameLength is in code points unit.
+-- while Range might not in code points unit.
+-- but we can still use it to get the length
+-- since it is only used to exclude some names
+-- currently, we only break `(ModuleA.b)` into `(ModuleA` and `.b)`
+splitModuleNameAndOccName :: VirtualFile -> Range -> Identifier -> [(Range,Identifier)]
+splitModuleNameAndOccName _ ran (Left m) = [(ran, Left m)]
+splitModuleNameAndOccName vf ran@(Range (Position startLine startColumn) (Position _endLine endColumn)) (Right name)
+    | nameLength name < fromIntegral (endColumn - startColumn), (Just text) <- getTextByCodePointRangeFromVfs vf ran =
```
Comment (michaelpj) [inline plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Utils.hs:114]: I'm still confused about why we can't use the text from the `OccName`? In particular, doesn't `occNameString . nameOccName` give you what you want? You even use that below!

Comment (michaelpj) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Utils.hs:114]: What I would kind of expect is something like:
- Check if the text is delimted by parens or backticks, remember this
- Match the name text against a regex to pull out the module prefix if there is one
- Compute the ranges based on the previous things:
   - If delimited, start and end one character earlier
   - If there is a module prefix, add a token for that bit, and start the operator bit later

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Utils.hs [hunk 6]
```diff
@@ -99,3 +101,63 @@ showSpan x = show (srcSpanStartLine x) <> ":" <> show (srcSpanStartCol x) <> "-"
...
 mkRange startLine startCol len =
     Range (Position (fromIntegral startLine) (fromIntegral startCol)) (Position (fromIntegral startLine) (fromIntegral $ startCol + len))
+
+
+-- nameLength is in code points unit.
+-- while Range might not in code points unit.
+-- but we can still use it to get the length
+-- since it is only used to exclude some names
+-- currently, we only break `(ModuleA.b)` into `(ModuleA` and `.b)`
+splitModuleNameAndOccName :: VirtualFile -> Range -> Identifier -> [(Range,Identifier)]
```
Comment (michaelpj) [inline plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Utils.hs:111]: This could do with a longer comment explaining what it's doing.

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Utils.hs [hunk 7]
```diff
@@ -99,3 +101,63 @@ showSpan x = show (srcSpanStartLine x) <> ":" <> show (srcSpanStartCol x) <> "-"
 mkRange :: (Integral a1, Integral a2) => a1 -> a2 -> a2 -> Range
 mkRange startLine startCol len =
     Range (Position (fromIntegral startLine) (fromIntegral startCol)) (Position (fromIntegral startLine) (fromIntegral $ startCol + len))
+
+
+-- nameLength is in code points unit.
+-- while Range might not in code points unit.
+-- but we can still use it to get the length
+-- since it is only used to exclude some names
+-- currently, we only break `(ModuleA.b)` into `(ModuleA` and `.b)`
```
Comment (michaelpj) [inline plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Utils.hs:110]: comment is wrong

File: plugins/hls-semantic-tokens-plugin/test/testdata/T1.expected [hunk 8]
```diff
@@ -72,7 +73,8 @@
 41:1-3 TFunction "go"
 41:6-9 TRecordField "foo"
 42:1-4 TFunction "add"
-42:7-18 TClassMethod "(Prelude.+)"
+42:8-16 TModuleName "Prelude."
```
Comment (michaelpj) [inline plugins/hls-semantic-tokens-plugin/test/testdata/T1.expected:76]: Can we have a test with a multiple-segment module name too? i.e. `(Foo.Bar.+)` or something

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Utils.hs (hunk 5 above)
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Utils.hs]: I update the doc, should be clear now.

File: plugins/hls-semantic-tokens-plugin/test/testdata/T1.expected (hunk 8 above)
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/test/testdata/T1.expected:76]: improve in TQualifiedName.expected

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs (hunk 1 above)
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs:16]: wait until the prioerities properties added.

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Utils.hs (hunk 5 above)
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Utils.hs]: > I'm still confused about why we can't use the text from the OccName? In particular, doesn't occNameString . nameOccName give you what you want? You even use that below!
the reason is here https://github.com/haskell/haskell-language-server/issues/3957#issuecomment-1893485590

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Query.hs (hunk 2 above)
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Query.hs]: done

**Comment by soulomoon** [comment]
@michaelpj  Thanx for the review. Should be good now.

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Query.hs [hunk 9]
```diff
@@ -62,39 +70,45 @@ nameNameSemanticFromHie hieKind rm ns = do
...
+        xs = S.fromList $ M.keys $ M.filterWithKey inclusion bd
         inclusion :: Identifier -> IdentifierDetails a -> Bool
         inclusion a b = not $ exclusion a b
         exclusion :: Identifier -> IdentifierDetails a -> Bool
         exclusion idt IdentifierDetails {identInfo = infos} = case idt of
-          Left _  -> True
-          Right _ -> any isEvidenceContext (S.toList infos)
+          Left _moduleName -> False
+          -- filter out the evidence names since they are visible,
+          -- derived names are not filtered out since they are not visible at use site.
```
Comment (michaelpj) [inline plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Query.hs:100]: Not the point of this PR, but this makes me wonder: won't such identifiers have a missing or zero-length span? So would it be enough to filter them when we create tokens by ignoring anything that has bad spans?

Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Query.hs:100]: Zero span sound like a ghc error.

Comment (michaelpj) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Query.hs:100]: what spans do the evidence names have?

**Comment by soulomoon** [comment]
> I'm also concerned that this is going to have a performance impact. We now query the VFS and do text processing for many (every?) token, which is a lot of additional work compared to just using the pre-computed spans in the HIE AST.
 
 yes, this is what I concerned too.
But this should be more ergonomic than the conversion we have done to support code points(this one is every token).
Since we have skipped a lot of case by comparing the source span and length of the occName

**Comment by soulomoon** [comment]
> I'd like a comment from @wz1000 about whether there is another way to get the full name text before we merge.

It would be great if there is way to get the source text for token more efficient.

**Comment by wz1000** [comment]
It does not seem ideal to do range lookups into the source text repeatedly. A better approach would be to walk the lexed source and the hie ast simultaneously, like the haddock hyperlinker:

https://gitlab.haskell.org/ghc/haddock/-/blob/ghc-head/haddock-api/src/Haddock/Backends/Hyperlinker/Renderer.hs?ref_type=heads#L78

**Comment by michaelpj** [comment]
It might also be fine if we can manage to only do this for things we know are either a) functions used infix, or b) operator sections. Can we recognize that from the HIE AST?

**Comment by soulomoon** [comment]
> It might also be fine if we can manage to only do this for things we know are either a) functions used infix, or b) operator sections. Can we recognize that from the HIE AST?

we can, we just check occname length and its span. If not equal, it is either: function used infix, operator section, qualified names.

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Query.hs (hunk 9 above)
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Query.hs:100]: the span it binds the evidence on

**Comment by soulomoon** [comment]
Walk along with the source code similar to https://github.com/haskell/haskell-language-server/pull/3958#issuecomment-1898180693

**Comment by wz1000** [comment]
I will review this over the weekend.

**Comment by michaelpj** [comment]
BTW @soulomoon you should join the matrix channel

**Comment by soulomoon** [comment]
> BTW @soulomoon you should join the matrix channel

sure, is it the address ? https://app.element.io/#/room/#haskell-tooling:matrix.org

**Comment by wz1000** [comment]
https://matrix.to/#/#haskell-language-server:matrix.org

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Types.hs [hunk 10]
```diff
@@ -25,6 +25,8 @@ import           Development.IDE.Graph.Classes (Hashable)
 import           GHC.Generics                  (Generic)
 import           Language.LSP.Protocol.Types
 -- import template haskell
+import           Data.Map                      (Map)
+import           Data.Set                      (Set)
```
Comment (wz1000) [inline plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Types.hs:29]: strict maps

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+          occStr <- lift $ case (occNameString . nameOccName) name of
+            -- the generated selector name with {-# LANGUAGE DuplicateRecordFields #-}
+            '$' : 's' : 'e' : 'l' : ':' : xs -> Just $ takeWhile (/= ':') xs
+            ['$']                            -> Just "$"
+            -- other generated names that should not be visible
+            '$' : _                          -> Nothing
+            ns                               -> Just ns
+          case ranSplit of
+            (NoSplit (tk, r)) -> do
+              guard $ T.unpack tk == occStr
```
Comment (wz1000) [inline plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs:119]: string comparisions can be expensive, using something like `Development.IDE.Types.Export.renderOcc = decodeUtf8 . bytesFS . occNameFS`, or even something like `encodeUtf8 text == bytesFS (occNameFS)` since encoding UTF8 with text 2.0+ is just a simple copy. Ideally we could avoid even the copy if we just compared the raw bytes in the `Text` and the `FastString` from `occNameFS` but this is not very easy to do.

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+  let span = nodeSpan leaf
+  (ran, token) <- focusTokenAt leaf
+  -- if `focusTokenAt` succeed, we can safely assume we have shift the cursor correctly
+  -- we do not need to recover the cursor state, even if the following computation failed
+  liftMaybeM $ do
+    -- only handle the leaf node with single column token
+    guard $ srcSpanStartLine span == srcSpanEndLine span
+    splitResult <-  lift $ splitRangeByText token ran
+    modify $ \s -> s {currentRange = ran, currentRangeContext = splitResult}
+    mapM_ combineNodeIds $ Map.filterWithKey (\k _ -> k == SourceInfo) $ getSourcedNodeInfo $ sourcedNodeInfo leaf
```
Comment (wz1000) [inline plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs:99]: Could we remove `currentRange` and `currentRangeContext` from the state if we passed as arguments to `combineNodeIds`?

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+  let (startPos, endPos) = srcSpanCharPositions span
+  startOff <- lift $ startPos `sub` cursor
+  tokenOff <- lift $ endPos `sub` startPos
+  (gap, startRope) <- lift $ charSplitAtPositionMaybe startOff rope
+  (token, remains) <- lift $ charSplitAtPositionMaybe tokenOff startRope
+  let ncs = newColumn columnsInUtf16 gap
+  let nce = newColumn ncs token
+  -- compute the new range for utf16, tuning the columns is enough
+  let ran = codePointRangeToRangeWith ncs nce $ realSrcSpanToCodePointRange span
+  modify $ \s -> s {columnsInUtf16 = nce, rope = remains, cursor = endPos}
```
Comment (wz1000) [inline plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs:143]: This stretch of code is a bit hard to follow. Perhaps it could use more comments explaining what is going on?

Here is my attempt:

We start with a `rope` representing the document remaining to be processed. `cursor` is a position (utf8?) relative to the start of the document, and is relative to the entire document, not the fragment we have in the `rope`. We want to focus on the position given by the `nodeSpan` of the `HieAST`, and return the token `(Range, Text)` at that span, advancing the `cursor` and the `rope` past the span. `nodeSpan`(utf8?) is also relative to the start of the entire document and not the fragment. `columnsInUtf16` is the number of columns in the utf8 encoding of the current line.

To obtain offsets corresponding to the `nodeSpan` in the fragment, we must subtract the `cursor` from the start position of the `nodeSpan`. This gives us a (utf8?) offset into `rope`. We subtract the start position from the endposition to obtain the (utf8?) length of the token. We can then advance the `cursor` and the `rope` past the token, obtaining the (utf8?) text of the current token in the process.

To obtain the utf16 range, we must advance the current columns first past the initial segment of the rope that we skip past (using its utf16 length, and accounting for newlines by only using the length from the last newline to the end of the segment), and then again past the current token (similarly). The first skip gives us the start column in utf16 bytes and the second skip gives us the end column, also in utf16 bytes. We finally update the `columnsInUtf16` using the end column.

**Review by wz1000 — CHANGES_REQUESTED** [review]
Looks good, just a few minor strictness and performance wiggles and a request for a bit more documentation.

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+{-# LANGUAGE RankNTypes          #-}
+
+module Ide.Plugin.SemanticTokens.Tokenize (hieAstSpanIdentifiers) where
+
+import           Control.Lens                    (Identity (runIdentity))
+import           Control.Monad                   (forM_, guard)
+import           Control.Monad.State             (MonadState (get),
+                                                  MonadTrans (lift), execStateT,
+                                                  gets, modify, put)
+import           Control.Monad.Trans.State       (StateT)
```
[the remark this answers is not in the document]
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs:14]: done

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Types.hs (hunk 10 above)
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Types.hs:29]: done

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+import           Ide.Plugin.SemanticTokens.Types (RangeIdSetMap)
+import           Language.LSP.Protocol.Types     (Position (Position),
+                                                  Range (Range), UInt, mkRange)
+import           Language.LSP.VFS                hiding (line)
+import           Prelude                         hiding (length, span)
+
+type Tokenizer m a = forall t. StateT (PTokenState t) m a
+
+
+data PTokenState t = PTokenState
```
[the remark this answers is not in the document]
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs]: done

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Query.hs [hunk 16]
```diff
@@ -1,100 +1,71 @@
...
 {-# LANGUAGE ScopedTypeVariables #-}
 
 -- |
 -- The query module is used to query the semantic tokens from the AST
 module Ide.Plugin.SemanticTokens.Query where
 
-import           Data.Either                          (rights)
 import           Data.Foldable                        (fold)
 import qualified Data.Map                             as M
 import qualified Data.Map                             as Map
```
[the remark this answers is not in the document]
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Query.hs:13]: done

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs [hunk 17]
```diff
@@ -21,8 +23,10 @@ import           Control.Monad.Except                     (ExceptT, liftEither,
                                                            withExceptT)
 import           Control.Monad.Trans                      (lift)
 import           Control.Monad.Trans.Except               (runExceptT)
-import           Data.Aeson                               (ToJSON (toJSON))
+import           Data.Map                                 (Map)
+import qualified Data.Map                                 as M
```
[the remark this answers is not in the document]
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs:27]: done

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs [hunk 18]
```diff
@@ -101,27 +106,28 @@ getSemanticTokensRule recorder =
...
-    let names = nameSetElemsStable $ unionNameSets $ Map.elems spanNamesMap
-    let localSemanticMap = mkLocalNameSemanticFromAst names (hieKindFunMasksKind hieKind) refMap
+    let spanIdMap = M.filter (not . null) $ hieAstSpanIdentifiers virtualFile ast
+    let names = S.toList $ S.unions $ Map.elems spanIdMap
+    let localSemanticMap = mkLocalIdSemanticFromAst names (hieKindFunMasksKind hieKind) refMap
     -- get imported name semantic map
-    let importedNameSemanticMap = foldr (getTypeExclude localSemanticMap getTyThingMap) emptyNameEnv names
-    let sMap = plusNameEnv_C (<>) importedNameSemanticMap localSemanticMap
-    let rangeTokenType = extractSemanticTokensFromNames sMap spanNamesMap
+    let importedIdSemanticMap = foldr (getTypeExclude localSemanticMap getTyThingMap) mempty names
```
[the remark this answers is not in the document]
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs:113]: good idea, done

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs (hunk 13 above)
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs]: yes, you are right, some renaming and doc are added now to make the code more readable

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs (hunk 18 above)
[the remark this answers is not in the document]
Comment (wz1000) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs:113]: Did you still mean to use `foldr` here? Why not `fromSet`?

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+      case idt of
+        Left _moduleName -> addRangeIdSetMap ran idt
+        Right name -> do
+          ranSplit <- gets currentRangeContext
+          occStr <- lift $ case (occNameString . nameOccName) name of
+            -- the generated selector name with {-# LANGUAGE DuplicateRecordFields #-}
+            '$' : 's' : 'e' : 'l' : ':' : xs -> Just $ takeWhile (/= ':') xs
+            ['$']                            -> Just "$"
+            -- other generated names that should not be visible
+            '$' : _                          -> Nothing
```
[the remark this answers is not in the document]
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs]: lets handle those as well to be sure.

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs (hunk 11 above)
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs]: revised, see if current appoach would be better.

**Comment by soulomoon** [comment]
@wz1000 Thank you for your detailed reviews. I have made some adjustment to the code.

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs (hunk 13 above)
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs]: Some comments are added to PTokenState
and Some added to this stretch of the code.

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs (hunk 18 above)
[the remark this answers is not in the document]
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs]: I was under the impression using `fromSet` would still have the same time complexity since I was using `elem` in getTypeExclude.

Now, giving a second thought, combining `fromSet` with `Set.difference` does reduce time complexity a bit for some cases.  reviced!

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs (hunk 11 above)
Comment (wz1000) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs]: I would try `encodeUtf8 text == bytesFS (occNameFS occ)` as that skips UTF-8 validation.

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+      case idt of
+        Left _moduleName -> addRangeIdSetMap ran idt
+        Right name -> do
+          occStr <- lift $ T.pack <$> case (occNameString . nameOccName) name of
+            -- the generated selector name with {-# LANGUAGE DuplicateRecordFields #-}
+            '$' : 's' : 'e' : 'l' : ':' : xs -> Just $ takeWhile (/= ':') xs
+            ['$']                            -> Just "$"
+            -- other generated names that should not be visible
+            '$' : _                          -> Nothing
+            _c : ':' : _                     -> Nothing
```
Comment (wz1000) [inline plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs:107]: I think you need a `isAlpha _c` guard.

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs (hunk 11 above)
Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs]: We are comparing the extracted text(`Text`) from rope with the occStr(`Text`) that have stripped the generated prefix.
I am not sure it is going to be better if we `encodeUtf8` on both side and compares them.🤔

File: plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+    getIdentifier ran ranSplit idt = liftMaybeM $ do
+      case idt of
+        Left _moduleName -> addRangeIdSetMap ran idt
+        Right name -> do
+          occStr <- lift $ T.pack <$> case (occNameString . nameOccName) name of
+            -- the generated selector name with {-# LANGUAGE DuplicateRecordFields #-}
+            '$' : 's' : 'e' : 'l' : ':' : xs -> Just $ takeWhile (/= ':') xs
+            ['$']                            -> Just "$"
+            -- other generated names that should not be visible
+            '$' : _                          -> Nothing
```
Comment (wz1000) [inline plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs:108]: You need the guard here as well I think

Comment (soulomoon) [reply plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs]: True, and added here and the previous case.

**Review by wz1000 — APPROVED** [review]

**Comment by soulomoon** [comment]
> another thing, though this doesn't have to be addressed in this MR and could be a follow up - now that we are traversing the entire document in order, do we still need to accumulate results into a `M.Map Range HsSemanticTokenType`? Couldn't we instead directly accumulate a `[(Range, HsSemanticTokenType)]`?

Good point, I'll take a look some time later if there is optimization space we can do on this.

**Comment by soulomoon** [comment]
> another thing, though this doesn't have to be addressed in this MR and could be a follow up - now that we are traversing the entire document in order, do we still need to accumulate results into a `M.Map Range HsSemanticTokenType`? Couldn't we instead directly accumulate a `[(Range, HsSemanticTokenType)]`?

One thing I am not sure about would be wether the mappings of range change would change the range ordering.
If the ordered would change, we are forced to sort it again.

**Comment by wz1000** [comment]
> One thing I am not sure about would be wether the mappings of range change would change the range ordering. If the ordered would change, we are forced to sort it again.

you mean the `positionMappings`? I'm pretty sure that they are guaranteed to be monotonic, so I don't think they would change sort order.

**Comment by soulomoon** [comment]
Yes, thanx for clearing this out!
